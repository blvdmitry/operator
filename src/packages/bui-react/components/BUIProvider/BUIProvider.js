import React from "react";
import { ExperimentProvider } from "../../hooks/useExperiment.js";
import ThemeProvider from "../ThemeProvider/index.js";
import { ToastProvider } from "../Toast/index.js";
import useGlobalRTL from "./hooks/useGlobalRTL.js";
import useGlobalKeyboardMode from "./hooks/useGlobalKeyboardMode.js";
import BUIContext from "./BUIProvider.context.js";
import "@bookingcom/bui-core/css/BUIProvider.css";
import "@bookingcom/bui-core/css/Mixin.module.css";
const BUIProvider = (props) => {
    const { children, defaultRTL, defaultViewportSize = "small", defaultThemeMode = "light", experiments = {}, theme, themeMode: passedThemeMode, } = props;
    const [rtl, setRTL] = useGlobalRTL(defaultRTL);
    const [themeMode, setThemeMode] = React.useState(defaultThemeMode);
    const value = React.useMemo(() => ({
        rtl,
        setRTL,
        defaultViewportSize,
        themeMode: passedThemeMode || themeMode,
        setThemeMode,
        invertThemeMode: () => {
            setThemeMode(themeMode === "light" ? "dark" : "light");
        },
    }), [rtl, setRTL, defaultViewportSize, themeMode, passedThemeMode, setThemeMode]);
    useGlobalKeyboardMode();
    const content = React.createElement(ToastProvider, null, children);
    return (React.createElement(ExperimentProvider, { value: experiments },
        React.createElement(BUIContext.Provider, { value: value }, theme ? (React.createElement(ThemeProvider, { theme: theme }, content)) : (content))));
};
export default BUIProvider;
