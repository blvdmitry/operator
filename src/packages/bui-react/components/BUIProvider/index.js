"use client";
export { default } from "./BUIProvider.js";
