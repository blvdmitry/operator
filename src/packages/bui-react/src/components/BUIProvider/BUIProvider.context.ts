import React from "react";
import type * as T from "./BUIProvider.types";

const BUIContext = React.createContext<T.Context>({
  rtl: false,
  setRTL: () => {},
  defaultViewportSize: "small",
  themeMode: "light",
  providerCount: 0,
  providerId: "",
  idRef: { current: 0 },
  setThemeMode: () => {},
  invertThemeMode: () => {},
});

export default BUIContext;
