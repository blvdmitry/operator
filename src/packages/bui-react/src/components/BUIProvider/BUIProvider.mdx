BUIProvider provides global meta information, like content direction and serves as a global wrapper for the whole application.

```
import { BUIProvider } from '@bookingcom/bui-react';
import LegacyTheme from '@bookingcom/bui-react/themes/legacy';

const App = ({ children }) => (
  <BUIProvider theme={LegacyTheme}>
    { children }
  </BUIProvider>
);
```

Anywhere inside the provider - you can get access to the BUI React hooks.

### Properties

| Name                | Type                                 | Default | Required | Description                              |
| ------------------- | ------------------------------------ | ------- | -------- | ---------------------------------------- |
| children            | `ReactNode`                          | -       | false    | Provider children elements               |
| theme               | `Theme`                              | -       | false    | Theme component                          |
| themeMode           | `light`, `dark`                      | false   |          | Controlled color mod                     |
| id                  | `string`                             | -       | false    |                                          |
| defaultThemeMode    | `light`, `dark`                      | -       | false    | Default theming mode for the application |
| defaultRTL          | `boolean`                            | -       | false    | Default RTL flag value for SSR           |
| defaultViewportSize | `small`, `medium`, `large`, `xlarge` | `small` | false    | Default Viewport Size for SSR            |
