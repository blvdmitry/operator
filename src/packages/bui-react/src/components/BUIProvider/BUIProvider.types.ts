import React from "react";
import type { ViewportName } from "@bookingcom/bui-core/types";
import type { ThemeProviderProps } from "components/ThemeProvider";
import type { Mode } from "types/theme";
import type { ExperimentName } from "types/experiments";

export type Context = {
  rtl: boolean;
  setRTL: (isRTL: boolean) => void;
  defaultViewportSize: ViewportName;
  themeMode: Mode;
  providerCount: number;
  providerId?: string;
  idRef: React.MutableRefObject<number>;
  setThemeMode: (mode: Mode) => void;
  invertThemeMode: () => void;
};

export type Props = {
  defaultRTL?: boolean;
  defaultViewportSize?: ViewportName;
  defaultThemeMode?: Mode;
  themeMode?: Mode;
  // Theme is optional here while Transport is passing their design tokens from their custom theming definitions
  theme?: ThemeProviderProps["theme"];
  // id of the provider for the case when multiple BUIProviders are used on the page coming from different package versions
  id?: string;
  experiments?: Partial<Record<ExperimentName, () => boolean>>;
  children?: React.ReactNode;
};
