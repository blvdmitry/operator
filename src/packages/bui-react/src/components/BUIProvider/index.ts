"use client";

export { default } from "./BUIProvider";
export type { Props as BUIProviderProps } from "./BUIProvider.types";
