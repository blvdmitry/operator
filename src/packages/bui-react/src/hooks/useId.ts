"use client";

import React from "react";
import BUIContext from "components/BUIProvider/BUIProvider.context";

const reactUseIdAvailable = typeof React.useId === "function";

export const useGetNextId = () => {
  const { idRef, providerCount, providerId } = React.useContext(BUIContext);

  return React.useCallback(() => {
    let contextPrefix = providerCount > 0 ? `c${providerCount}-` : "";
    if (providerId) contextPrefix = `${providerId}-${contextPrefix}`;
    return `__bui-${contextPrefix}${++idRef.current}`;
  }, [idRef, providerCount, providerId]);
};

const useId = (id?: string): string => {
  const getNextId = useGetNextId();

  if (id !== undefined) return id;

  if (reactUseIdAvailable) {
    // eslint-disable-next-line react-hooks/rules-of-hooks
    return React.useId();
  }

  // eslint-disable-next-line react-hooks/rules-of-hooks
  return React.useMemo(() => {
    if (id) return id;
    return getNextId();
  }, [getNextId, id]);
};

export default useId;
