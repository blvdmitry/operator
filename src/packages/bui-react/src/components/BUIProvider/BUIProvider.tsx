import React from "react";
import { ExperimentProvider } from "hooks/useExperiment";
import ThemeProvider from "components/ThemeProvider";
import useGlobalRTL from "./hooks/useGlobalRTL";
import useGlobalKeyboardMode from "./hooks/useGlobalKeyboardMode";
import BUIContext from "./BUIProvider.context";
import type * as T from "./BUIProvider.types";
import "@bookingcom/bui-core/css/BUIProvider.css";
import "@bookingcom/bui-core/css/Mixin.module.css";

let providerCount = 0;

const BUIProvider = (props: T.Props) => {
  const {
    children,
    defaultRTL,
    defaultViewportSize = "small",
    defaultThemeMode = "light",
    experiments = {},
    theme,
    themeMode: passedThemeMode,
    id,
  } = props;
  const [rtl, setRTL] = useGlobalRTL(defaultRTL);
  const [themeMode, setThemeMode] = React.useState(defaultThemeMode);
  const idRef = React.useRef(0);
  // Temporary solution to make id generation unique across multiple BUI providers
  // To be removed once we migrate to React 18 and start using a new hook provided by React
  // https://github.com/facebook/react/pull/22644
  const count = React.useMemo(() => {
    const prefix = providerCount;
    providerCount++;
    return prefix;
  }, []);

  const value = React.useMemo(
    () => ({
      rtl,
      setRTL,
      defaultViewportSize,
      themeMode: passedThemeMode || themeMode,
      setThemeMode,
      idRef,
      providerCount: count,
      providerId: id,
      invertThemeMode: () => {
        setThemeMode(themeMode === "light" ? "dark" : "light");
      },
    }),
    [
      rtl,
      setRTL,
      defaultViewportSize,
      themeMode,
      passedThemeMode,
      setThemeMode,
      idRef,
      count,
      id,
    ]
  );

  useGlobalKeyboardMode();

  return (
    <ExperimentProvider value={experiments}>
      <BUIContext.Provider value={value}>
        {theme ? (
          <ThemeProvider theme={theme}>{children}</ThemeProvider>
        ) : (
          children
        )}
      </BUIContext.Provider>
    </ExperimentProvider>
  );
};

export default BUIProvider;
