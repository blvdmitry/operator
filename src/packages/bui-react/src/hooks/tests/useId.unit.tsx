import React from "react";
import { render, screen } from "@testing-library/react";
import BUIProvider from "components/BUIProvider";
import useId from "hooks/useId";
import TravellerTheme from "themes/traveller";
import mockedUseId from "./__mocks__/useId";

jest.mock("../useId.ts", () => ({
  __esModule: true,
  default: jest.fn((id) => mockedUseId(id)),
}));

const Generator = () => {
  const id = useId();

  return <div data-testid="id">{id}</div>;
};

describe("useId", () => {
  test("generates ids", () => {
    render(
      <BUIProvider theme={TravellerTheme}>
        <Generator />
        <Generator />
        <Generator />
      </BUIProvider>
    );

    const ids = screen.getAllByTestId("id");
    expect(ids).toMatchSnapshot();
  });

  test("works with multiple providers", () => {
    render(
      <>
        <BUIProvider theme={TravellerTheme}>
          <Generator />
          <Generator />
          <Generator />
        </BUIProvider>
        <BUIProvider theme={TravellerTheme}>
          <Generator />
          <Generator />
          <Generator />
        </BUIProvider>
      </>
    );

    const ids = screen.getAllByTestId("id");
    expect(ids).toMatchSnapshot();
  });
});
