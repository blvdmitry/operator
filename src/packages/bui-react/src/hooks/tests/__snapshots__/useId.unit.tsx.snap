// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useId generates ids 1`] = `
[
  <div
    data-testid="id"
  >
    __bui-1
  </div>,
  <div
    data-testid="id"
  >
    __bui-2
  </div>,
  <div
    data-testid="id"
  >
    __bui-3
  </div>,
]
`;

exports[`useId works with multiple providers 1`] = `
[
  <div
    data-testid="id"
  >
    __bui-c1-1
  </div>,
  <div
    data-testid="id"
  >
    __bui-c1-2
  </div>,
  <div
    data-testid="id"
  >
    __bui-c1-3
  </div>,
  <div
    data-testid="id"
  >
    __bui-c2-1
  </div>,
  <div
    data-testid="id"
  >
    __bui-c2-2
  </div>,
  <div
    data-testid="id"
  >
    __bui-c2-3
  </div>,
]
`;
